core/src/p2p.rs

```diff
@@ -0,0 +1,741 @@
+                    .any(|key| key == &peer_id.public_key)
+            })
+            .map(|peer_id| peer_id.address.clone())
+            .collect();
+        if target_addrs.is_empty() {
+            None
+        } else {
+            let address_candidate = &target_addrs[rand::random::<usize>() % target_addrs.len()];
+            // to_socket_addrs enables dns look-ups.
+            let maybe_addr = address_candidate
+                .to_socket_addrs()
+                .unwrap_or_else(|error| {
+                    debug!(?error, "Socket addr");
+                    Vec::new().into_iter()
+                })
+                .next();
+            if maybe_addr.is_none() {
+                error!(%address_candidate, "Error can't produce address from str. ");
+            }
+            maybe_addr
+        }
+    }
+
+    pub(crate) fn post_to_network(&self, message: &NetworkMessage, recipients: &[PublicKey]) {
+        let mut to_disconnect_keys = Vec::new();
+
+        let mut connected_to_peers = self
+            .connected_to_peers
+            .lock();
+        for public_key in recipients.iter() {
+            if let Some(PeerConnection { stream, crypto, .. }) =
+                connected_to_peers.get_mut(public_key)
+            {
+                let encrypted = crypto
+                    .encrypt(message.encode().clone())
+                    .expect("We should always be able to encrypt.");
+                let write1 = stream.write_all(&(u32::try_from(encrypted.len())).expect("Encrypted length exceeds 32 bits. This will cause issues with Parity Scale codec. Aborting. ").to_le_bytes());
+                let write2 = stream.write_all(&encrypted);
+
+                if write1.is_err() || write2.is_err() {
+                    error!(
+                        "Disconecting, cause: Could not post message to peer {}.",
+                        public_key
+                    );
+                    to_disconnect_keys.push(public_key.clone());
+                }
+            } else {
+                debug!("P2P post failed: Not connected.");
+            }
+        }
+        drop(connected_to_peers);
+        self.disconnect_peers(to_disconnect_keys);
+    }
+
+    #[allow(
+        clippy::unwrap_in_result,
+        clippy::redundant_else,
+        clippy::cognitive_complexity,
+        clippy::needless_pass_by_value
+    )]
+    pub(crate) fn poll_network_for_packet(
+        &self,
+        variant: NetworkMessageVariant,
+    ) -> Option<NetworkMessage> {
+        let ReadThreadData {
+            ref mut poll_network_index,
+            ref mut packet_cache,
+        } = *self
+            .read_thread_data
+            .lock();
+
+        let mut insert_index = 0;
+        for i in 0..packet_cache.len() {
+            if let Some(content) = packet_cache[i].take() {
+                match content {
+                    SumeragiPacket(internal) => {
+                        if matches!(variant, NetworkMessageVariant::Sumeragi) {
```

Comment:
Rewrite using an `if let`. This does the exact same as `matches!`, which is introduced to help writing `assert!` statements and not for general use.